keps/sig-network/0031-20181109-multicast-support.md

```diff
@@ -0,0 +1,191 @@
+* [Proposal](#proposal)
+    * [User Stories](#user-stories)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+Currently, support for IP multicast traffic on the pod network is
+entirely undefined, with no multicast-related behavior being either
+required or forbidden, and different plugins offering different
+functionality. This KEP aims to standardize things.
+
+## Motivation
+
+Some Kubernetes users have workloads that require multicast traffic,
+and need to be able to confirm that multicast traffic will flow
+between the desired endpoints, but not be visible to other endpoints.
+
+Other Kubernetes users are more concerned with simply ensuring that
+undesired multicast traffic is blocked, to prevent a single pod from
+being able to easily saturate the network.
+
+See also:
+[Multicast in Weave](https://www.weave.works/blog/multicasting-cloud-need-know/),
+[Multicast in OpenShift](https://blog.openshift.com/service-discovery-openshift-using-multicast/).
+
+### Goals
+
+- Explicitly define the expected baseline behavior of multicast traffic in a Kubernetes cluster.
+- Define additional levels of network plugin multicast support beyond the baseline.
+- Define how users enable/disable/configure multicast support and how it interacts with NetworkPolicy.
+
+### Non-Goals
+
+- Requiring all plugins to support multicast.
+- Anything having to do with multicast on interfaces/networks other than the default cluster network.
+
+## Existing Work
+
+Most plugins currently block all multicast (either intentionally, or
+just accidentally as a consequence of IP/MAC filtering).
+
+Weave treats multicast as broadcast. There has been discussion of
+[implementing IGMP
+snooping](https://github.com/weaveworks/weave/issues/178) so that
+multicast packets would only be sent to the pods that want them, but
+this is not yet implemented. There has been some discussion of
+[multicast between pods and external
+IPs](https://github.com/weaveworks/weave/issues/1863), which
+apparently can be made to work at least in some circumstances. There
+is no support for policy/isolation; all multicast packets are sent to
+all pods in the cluster.
+
+OpenShift SDN also implements multicast-as-broadcast, but on a
+per-namespace basis, [with namespaces having to opt
+in](https://docs.okd.io/latest/admin_guide/managing_networking.html#admin-guide-networking-multicast).
+(So if you enable multicast on namespace "foo", and a pod in "foo"
+sends out a multicast packet, it will be delivered to all other pods
+in "foo".) As with Weave, it's not possible to use NetworkPolicy to
+restrict multicast traffic, and it's also not possible to use
+NetworkPolicy to extend multicast traffic across namespace boundaries.
+
+## Proposal
+
+### User Stories
+
+TBD
```

Comment:
The JBoss service discovery use-case is probably the best defined one here at the moment.